### ISSUE TITLE
Cannot start Java process with Environment variables containing spaces

### ISSUE THREAD

**Issue by karoldrazek** [systemd/systemd#8136]
### Submission type

 - Bug report

### systemd version the issue has been seen with

systemd 219
+PAM +AUDIT +SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ -LZ4 -SECCOMP +BLKID +ELFUTILS +KMOD +IDN

### Used distribution

> Centos 7.3, Systemd package release 42.el7_4.7

### In case of bug report: Expected behaviour you didn't see

Using following script:
```
[Unit]
Description=knotx-main
After=network.target

[Service]
EnvironmentFile=/opt/knotx/main/knotx.conf
ExecStart=/usr/bin/java \
  -Dlogback.configurationFile=/opt/knotx/main/logback.xml \
  -Dvertx.cacheDirBase=/opt/knotx/main/.vertx \
  -Xms${KNOTX_MIN_HEAP}m \
  -Xmx${KNOTX_MAX_HEAP}m \
  -XX:ReservedCodeCacheSize=${KNOTX_CODE_CACHE}m \
  -XX:+UseBiasedLocking \
  -XX:BiasedLockingStartupDelay=0 \
  ${KNOTX_GENERAL_OPTS} ${KNOTX_JMX_OPTS} ${KNOTX_GC_OPTS} \
  ${KNOTX_DEBUG_OPTS} \
  ${KNOTX_EXTRA_OPTS} \
  -Dio.knotx.KnotxServer.options.config.serverOptions.port=${KNOTX_PORT} \
  -cp "/opt/knotx/main/app/*" io.knotx.launcher.LogbackLauncher \
  ${KNOTX_CONFIG} \
  ${KNOTX_CONFIG_EXTRA}
Type=simple
PIDFile=/var/run/knotx-main
User=knotx
LimitNOFILE=65536

[Install]
WantedBy=default.target
```

and EnvironmentFile:
```
# Knotx general attributes

KNOTX_CONFIG="-conf /opt/knotx/main/config.json"
KNOTX_CONFIG_EXTRA="-test"

# Knotx JVM attributes

KNOTX_MIN_HEAP=256
KNOTX_MAX_HEAP=1024
KNOTX_CODE_CACHE=64
KNOTX_PORT=8092
KNOTX_EXTRA_OPTS=-Dtest

KNOTX_GENERAL_OPTS="-Djava.net.preferIPv4Stack=true"

KNOTX_GC_OPTS="-XX:+UseConcMarkSweepGC -XX:ParallelGCThreads=2 -XX:ParallelCMSThreads=1 -Xloggc:/var/log/knotx/gc-main.log -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCApplicationConcurrentTime -XX:+PrintGC -XX:+PrintGCTimeStamps -XX:+PrintGCDetails -XX:+UseGCLogFileRotation -XX:NumberOfGCLogFiles=10 -XX:GCLogFileSize=5M"

KNOTX_JMX_OPTS="-Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=18092 -Dcom.sun.management.jmxremote.rmi.port=18092 -Djava.rmi.server.hostname=0.0.0.0 -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote.authenticate=false"

KNOTX_DEBUG_OPTS="-agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=28092"
```

### In case of bug report: Unexpected behaviour you saw

I'm getting following errors in log:
```
Feb 08 19:21:40 config-file-centos-73.vagrantup.com systemd[1]: Converting job knotx-main.service/restart -> knotx-main.service/start
Feb 08 19:21:40 config-file-centos-73.vagrantup.com systemd[1]: About to execute: /usr/bin/java -Dlogback.configurationFile=/opt/knotx/main/logback.xml -Dvertx.cacheDirBase=/opt/knotx/main/.vertx -Xms${KNOTX_MIN_HEAP}m -Xmx${KNOTX_MAX_HEAP}m -XX:ReservedCodeCacheSize=${KNOTX_CODE_CACHE}m -XX:+UseBiasedLocking -XX:BiasedLockingStartupDelay=0 ${KNOTX_GENERAL_OPTS} ${KNOTX_JMX_OPTS} ${KNOTX_GC_OPTS} ${KNOTX_DEBUG_OPTS} ${KNOTX_EXTRA_OPTS} -Dio.knotx.KnotxServer.options.config.serverOptions.port=${KNOTX_PORT} -cp /opt/knotx/main/app/* io.knotx.launcher.LogbackLauncher ${KNOTX_CONFIG} ${KNOTX_CONFIG_EXTRA}
Feb 08 19:21:40 config-file-centos-73.vagrantup.com systemd[1]: Forked /usr/bin/java as 9033
Feb 08 19:21:40 config-file-centos-73.vagrantup.com systemd[1]: knotx-main.service changed failed -> running
Feb 08 19:21:40 config-file-centos-73.vagrantup.com systemd[1]: Job knotx-main.service/start finished, result=done
Feb 08 19:21:40 config-file-centos-73.vagrantup.com systemd[1]: Started knotx-main.
Feb 08 19:21:40 config-file-centos-73.vagrantup.com systemd[1]: Starting knotx-main...
Feb 08 19:21:40 config-file-centos-73.vagrantup.com systemd[1]: Child 9033 belongs to knotx-main.service
Feb 08 19:21:40 config-file-centos-73.vagrantup.com systemd[1]: knotx-main.service: main process exited, code=exited, status=1/FAILURE
Feb 08 19:21:40 config-file-centos-73.vagrantup.com systemd[1]: knotx-main.service changed running -> failed
Feb 08 19:21:40 config-file-centos-73.vagrantup.com systemd[1]: Unit knotx-main.service entered failed state.
Feb 08 19:21:40 config-file-centos-73.vagrantup.com systemd[1]: knotx-main.service failed.
Feb 08 19:21:40 config-file-centos-73.vagrantup.com systemd[1]: knotx-main.service: cgroup is empty
Feb 08 19:21:47 config-file-centos-73.vagrantup.com systemd[1]: Started knotx-main.
Feb 08 19:21:47 config-file-centos-73.vagrantup.com systemd[1]: Starting knotx-main...
Feb 08 19:21:47 config-file-centos-73.vagrantup.com systemd[9063]: Executing: /usr/bin/java -Dlogback.configurationFile=/opt/knotx/main/logback.xml -Dvertx.cacheDirBase=/opt/knotx/main/.vertx -Xms256m -Xmx1024m -XX:ReservedCodeCacheSize=64m -XX:+UseBiasedLocking -XX:BiasedLockingStartupDelay=0 -Djava.net.preferIPv4Stack=true '-Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=18092 -Dcom.sun.management.jmxremote.rmi.port=18092 -Djava.rmi.server.hostname=0.0.0.0 -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote.authenticate=false' '-XX:+UseConcMarkSweepGC -XX:ParallelGCThreads=2 -XX:ParallelCMSThreads=1 -Xloggc:/var/log/knotx/gc-main.log -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCApplicationConcurrentTime -XX:+PrintGC -XX:+PrintGCTimeStamps -XX:+PrintGCDetails -XX:+UseGCLogFileRotation -XX:NumberOfGCLogFiles=10 -XX:GCLogFileSize=5M' -agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=28092 -Dtest -Dio.knotx.KnotxServer.options.config.serverOptions.port=8092 -cp /opt/knotx/main/app/* io.knotx.launcher.LogbackLauncher '-conf /opt/knotx/main/config.json' -test
Feb 08 19:21:47 config-file-centos-73.vagrantup.com java[9063]: Unrecognized VM option 'UseConcMarkSweepGC -XX:ParallelGCThreads=2 -XX:ParallelCMSThreads=1 -Xloggc:/var/log/knotx/gc-main.log -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCApplicationConcurrentTime -XX:+PrintGC -XX:+PrintGCTimeStamps -XX:+PrintGCDetails -XX:+UseGCLogFileRotation -XX:NumberOfGCLogFiles=10 -XX:GCLogFileSize=5M'
Feb 08 19:21:47 config-file-centos-73.vagrantup.com java[9063]: Error: Could not create the Java Virtual Machine.
Feb 08 19:21:47 config-file-centos-73.vagrantup.com java[9063]: Error: A fatal exception has occurred. Program will exit.
Feb 08 19:21:47 config-file-centos-73.vagrantup.com systemd[1]: knotx-main.service: main process exited, code=exited, status=1/FAILURE
Feb 08 19:21:47 config-file-centos-73.vagrantup.com systemd[1]: Unit knotx-main.service entered failed state.
Feb 08 19:21:47 config-file-centos-73.vagrantup.com systemd[1]: knotx-main.service failed.
```
I also tried adding space in front of garbage collection attributes:
```
KNOTX_GC_OPTS=" -XX:+UseConcMarkSweepGC -XX:ParallelGCThreads=2 -XX:ParallelCMSThreads=1 -Xloggc:/var/log/knotx/gc-main.log -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCApplicationConcurrentTime -XX:+PrintGC -XX:+PrintGCTimeStamps -XX:+PrintGCDetails -XX:+UseGCLogFileRotation -XX:NumberOfGCLogFiles=10 -XX:GCLogFileSize=5M"
```
That unfortunately also causes issues with unit start up:
```
Feb 08 19:25:50 config-file-centos-73.vagrantup.com systemd[1]: Converting job knotx-main.service/restart -> knotx-main.service/start
Feb 08 19:25:50 config-file-centos-73.vagrantup.com systemd[1]: About to execute: /usr/bin/java -Dlogback.configurationFile=/opt/knotx/main/logback.xml -Dvertx.cacheDirBase=/opt/knotx/main/.vertx -Xms${KNOTX_MIN_HEAP}m -Xmx${KNOTX_MAX_HEAP}m -XX:ReservedCodeCacheSize=${KNOTX_CODE_CACHE}m -XX:+UseBiasedLocking -XX:BiasedLockingStartupDelay=0 ${KNOTX_GENERAL_OPTS} ${KNOTX_JMX_OPTS} ${KNOTX_GC_OPTS} ${KNOTX_DEBUG_OPTS} ${KNOTX_EXTRA_OPTS} -Dio.knotx.KnotxServer.options.config.serverOptions.port=${KNOTX_PORT} -cp /opt/knotx/main/app/* io.knotx.launcher.LogbackLauncher ${KNOTX_CONFIG} ${KNOTX_CONFIG_EXTRA}
Feb 08 19:25:50 config-file-centos-73.vagrantup.com systemd[1]: Forked /usr/bin/java as 9095
Feb 08 19:25:50 config-file-centos-73.vagrantup.com systemd[1]: knotx-main.service changed failed -> running
Feb 08 19:25:50 config-file-centos-73.vagrantup.com systemd[1]: Job knotx-main.service/start finished, result=done
Feb 08 19:25:50 config-file-centos-73.vagrantup.com systemd[1]: Started knotx-main.
Feb 08 19:25:50 config-file-centos-73.vagrantup.com systemd[1]: Starting knotx-main...
Feb 08 19:25:50 config-file-centos-73.vagrantup.com systemd[9095]: Executing: /usr/bin/java -Dlogback.configurationFile=/opt/knotx/main/logback.xml -Dvertx.cacheDirBase=/opt/knotx/main/.vertx -Xms256m -Xmx1024m -XX:ReservedCodeCacheSize=64m -XX:+UseBiasedLocking -XX:BiasedLockingStartupDelay=0 -Djava.net.preferIPv4Stack=true '-Dcom.sun.management.jmxremote -Dcom.sun.management.jmxremote.port=18092 -Dcom.sun.management.jmxremote.rmi.port=18092 -Djava.rmi.server.hostname=0.0.0.0 -Dcom.sun.management.jmxremote.ssl=false -Dcom.sun.management.jmxremote.authenticate=false' ' -XX:+UseConcMarkSweepGC -XX:ParallelGCThreads=2 -XX:ParallelCMSThreads=1 -Xloggc:/var/log/knotx/gc-main.log -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCApplicationConcurrentTime -XX:+PrintGC -XX:+PrintGCTimeStamps -XX:+PrintGCDetails -XX:+UseGCLogFileRotation -XX:NumberOfGCLogFiles=10 -XX:GCLogFileSize=5M' -agentlib:jdwp=transport=dt_socket,server=y,suspend=n,address=28092 -Dtest -Dio.knotx.KnotxServer.options.config.serverOptions.port=8092 -cp /opt/knotx/main/app/* io.knotx.launcher.LogbackLauncher '-conf /opt/knotx/main/config.json' -test
Feb 08 19:25:50 config-file-centos-73.vagrantup.com java[9095]: Error: Could not find or load main class  -XX:+UseConcMarkSweepGC -XX:ParallelGCThreads=2 -XX:ParallelCMSThreads=1 -Xloggc:.var.log.knotx.gc-main.log -XX:+PrintGCApplicationStoppedTime -XX:+PrintGCApplicationConcurrentTime -XX:+PrintGC -XX:+PrintGCTimeStamps -XX:+PrintGCDetails -XX:+UseGCLogFileRotation -XX:NumberOfGCLogFiles=10 -XX:GCLogFileSize=5M
Feb 08 19:25:50 config-file-centos-73.vagrantup.com systemd[1]: Child 9095 belongs to knotx-main.service
Feb 08 19:25:50 config-file-centos-73.vagrantup.com systemd[1]: knotx-main.service: main process exited, code=exited, status=1/FAILURE
Feb 08 19:25:50 config-file-centos-73.vagrantup.com systemd[1]: knotx-main.service changed running -> failed
Feb 08 19:25:50 config-file-centos-73.vagrantup.com systemd[1]: Unit knotx-main.service entered failed state.
Feb 08 19:25:50 config-file-centos-73.vagrantup.com systemd[1]: knotx-main.service failed.
```

### In case of bug report: Steps to reproduce the problem

I've tested lots of scenarios and it seems the core issue here is that while reading variables where space sign is present, systemd adds single quotes in startup commands for that part. This causes Java base apps to fail.
I've checked both scenarios where variables are read from Environment and EnvironmentFile.

Right now it seems the only way to go with it is to put each ExecStart parameters directly into single variables and use those in ExecStart or just put the parameters directly into .service file.

**Comment by yuwata** [comment]
Could you test this with the recent release v237 or v236, as CONTRIBUTING.md saids
> We only track bugs in the two most recently released versions of systemd in the GitHub Issue tracker. If you are using an older version of systemd, please contact your distribution's bug tracker instead.

If you still have this in recent version, then please reopen this. Thank you.

**Comment by bl33pbl0p** [comment]
Even though this is an old version, and that it shouldn't be tracked here but in your downstream bugtracker, behaviour regarding variable expansion has remained unchanged in systemd, which is a bit different from everywhere else. The subtleties were documented in a later version, before which it was phrased in a manner that would make you think it is shell like. Quoting from 236,

> Basic environment variable substitution is supported. Use "${FOO}" as part of a word, or as a word of its own, on the command line, in which case it will be replaced by the value of the environment variable including all whitespace it contains, resulting in a single argument. Use "$FOO" as a separate word on the command line, in which case it will be replaced by the value of the environment variable split at whitespace, resulting in zero or more arguments. For this type of expansion, quotes are respected when splitting into words, and afterwards removed.

Example:

Environment="ONE=one" 'TWO=two two'
ExecStart=/bin/echo $ONE $TWO ${TWO}
This will execute /bin/echo with four arguments: "one", "two", "two", and "two two".

Example:

Environment=ONE='one' "TWO='two two' too" THREE=
ExecStart=/bin/echo ${ONE} ${TWO} ${THREE}
ExecStart=/bin/echo $ONE $TWO $THREE
This results in echo being called twice, the first time with arguments "'one'", "'two two' too", "", and the second time with arguments "one", "two two", "too".

In short, $FOO does word splitting and quote stripping, while ${FOO} does no word splitting and quote stripping. This is indeed a bit odd but is now documented correctly.
